java/ql/src/experimental/Security/CWE/CWE-094/JexlInjectionLib.qll

```diff
@@ -0,0 +1,231 @@
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking
+
+/**
+ * A taint-tracking configuration for unsafe user input
+ * that is used to construct and evaluate a Jexl expression.
+ * It supports both Jexl2 and Jexl3.
+ */
+class JexlInjectionConfig extends TaintTracking::Configuration {
+  JexlInjectionConfig() { this = "JexlInjectionConfig" }
+
+  override predicate isSource(DataFlow::Node source) {
+    source instanceof TaintedSpringRequestBody or
+    source instanceof RemoteFlowSource
+  }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof JexlEvaluationSink }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node fromNode, DataFlow::Node toNode) {
+    any(TaintPropagatingJexlMethodCall c).taintFlow(fromNode, toNode) or
+    returnsDataFromBean(fromNode, toNode)
+  }
+}
+
+/**
+ * A data flow source for parameters that have
+ * a Spring framework annotation indicating remote user input from servlets.
+ */
+private class TaintedSpringRequestBody extends DataFlow::Node {
+  TaintedSpringRequestBody() {
+    exists(SpringServletInputAnnotation a | this.asParameter().getAnAnnotation() = a)
+  }
+}
+
+/**
+ * A sink for Expresssion Language injection vulnerabilities via Jexl,
+ * i.e. method calls that run evaluation of a Jexl expression.
+ */
+private class JexlEvaluationSink extends DataFlow::ExprNode {
+  JexlEvaluationSink() {
+    exists(MethodAccess ma, Method m, Expr taintFrom |
+      ma.getMethod() = m and taintFrom = this.asExpr()
+    |
+      m instanceof DirectJexlEvaluationMethod and ma.getQualifier() = taintFrom
+      or
+      m instanceof CallableCallMethod and ma.getQualifier() = taintFrom
+      or
+      m instanceof JexlEngineGetSetPropertyMethod and
+      ma.getAnArgument().getType() instanceof TypeString and
+      ma.getAnArgument() = taintFrom
```

Comment:
I think you need to check that both arguments have the same index /cc @smowton

Reply:
Doh, yes, you could check `taintFrom.getType()` to make sure you're referring to the same arg

Reply:
I've updated the query, thanks!